### DISCUSSION TITLE
How to export .tiff with multiple Alpha channels.

### DISCUSSION THREAD

**Discussion by LuukOV** [ImageMagick/ImageMagick#5840] [Help] [answered]
Hello everyone, I've been trying to export a .tiff file to have multiple Alpha channels. I need 1 to be (un)associated and 1 to be unspecified. I am able to generate this through photoshop so I know it's possible, but I haven't been able to do it through code. How would I go about doing this?
My current code:
`exec('gimp -i -d -f -b '(let* ((img (car (file-png-load 1 \"example.png\" \"example.png\")))
     (drawable (car (gimp-image-active-drawable img))))
     (gimp-image-select-item img 0 drawable)
     (plug-in-sel2path 0 img drawable)
     (gimp-image-insert-channel img (car (gimp-channel-new-from-component img 5 \"Alpha 1\")) 0 5)
     (gimp-image-insert-channel img (car (gimp-channel-new-from-component img 5 \"Transparency\")) 0 6)
     (file-tiff-save 1 img drawable \"output.tiff\" \"output.tiff\" 1) (gimp-quit 0))'`, {cwd: `src/images/${r}', env:process.env}, async (error, stdout, stderr) => {                        
        im.convert(['${appDir}/images/${r}/output.tiff', "-define", 
        "tiff:alpha=unspecified", '${appDir}/images/${r}/output2.tiff'], () => {
            //script where I save the the output image
        })
})`

**Comment by snibgo** [comment]
What version of IM, on what platform?

I assume you also want three colour channels (R,G,B).

Images in ImageMagick have only one alpha channel, but can also have meta-channels. Here is one method to do what you want.

Write the pixel values to a text file that I name `fmt2.lis` like this:
```
0,0:0,10%,20%,30%,40%
0,1:0,11%,22%,33%,44%
```
Both lines must have a new-line, either "\n" (as usual for unix) or "\r\n" (as usual for Windows).

The file has x and y coordinates, a colon, then five values. This example has only two pixels, for an image that is size 1x2.

Read this with IM, and write a TIFF, like this:
```
magick -size 1x2 -define ftxt:hasalpha=true -define ftxt:nummeta=1 ftxt:fmt2.lis x.tiff
```
tiffinfo than says this about the TIFF file:
```
C:\web\im>tiffinfo x.tiff
=== TIFF directory 0 ===
TIFF Directory at offset 0x1c (28)
  Image Width: 1 Image Length: 2
  Bits/Sample: 16
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 2<unassoc-alpha, unspecified>
  FillOrder: msb-to-lsb
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 5
  Rows/Strip: 2
  Planar Configuration: single image plane
  Page Number: 0-1
  White Point: 0.3127-0.329
  PrimaryChromaticities: 0.640000,0.330000,0.300000,0.600000,0.150000,0.060000
```
As required, the file has 5 channels. 3 are for RGB, and the extra samples are "unassoc-alpha" and "unspecified".

IM can read the TIFF file, and write ftxt:
```
C:\web\im>%IMG7%magick x.tiff txt:
# ImageMagick pixel enumeration: 1,2,1,65535,srgba
0,0: (0,6554,13107,19661,26214)  #0000199A33334CCD  srgba(0%,10.0008%,20%,0.300008)
0,1: (0,7209,14418,21627,28835)  #00001C293852547B  srgba(0%,11.0002%,22.0005%,0.330007)
```
The values in parentheses are shown on a scale of 0 to QuantumRange.

For more details about "ftxt:", see [ftxt: formatted text](http://im.snibgo.com/fmttxt.htm).

**Reply by LuukOV** [reply]
Thank you! From what I understand, this generates a sample file which has the structure of channels that I would want. Which is already possibly a big step in the right direction. How, however, would I go about and applying this to an already existing .tiff/png which has 1 alpha channel. I assume I would have to duplicate the existing alpha channel and define this as the 5th channel somehow?

**Comment by snibgo** [comment]
> How, however, would I go about and applying this to an already existing .tiff/png which has 1 alpha channel. I assume I would have to duplicate the existing alpha channel and define this as the 5th channel somehow?

Where does the data for your fifth channel come from? Is it simply a copy of the alpha channel? That seems pointless. Doesn't that data come from a different image?

**Reply by LuukOV** [reply]
Generally it should just be a duplicate of the current alpha channel. My goals is to get a result similar to this, if not exactly this: 
[https://i.imgur.com/vQ93fRM.png](https://i.imgur.com/vQ93fRM.png)
Pay attention to the transparent background in the RGB channels, together with the existence of an Alpha channel.
If you export this within photoshop, it saves it as 2 alpha channels just like how I described it initially.

**Comment by snibgo** [comment]
Please answer my questions: what version of IM are you using? On what platform (bash, Windows, whatever)?

The screenshot you show is of an image that seems to have just one alpha channel.

Please link to a file that contains five channels, as you want them. Link to the actual file, not a screenshot of some program that is displaying the file.

It would help if you also linked to a sample input file.

**Reply by LuukOV** [reply]
IM: 0.1.3 (https://github.com/yourdeveloper/node-imagemagick) (NPM)
Being run on a Node.js server running on Debian. So I call the functions using JS.
An example of the result I want to achieve: 
https://we.tl/t-ACQuXtMX2u (it doesn't allow me to upload .tiff here)
An example of the file I use as input:
![input](https://user-images.githubusercontent.com/27200792/205287075-75d71763-5121-447b-a956-08c8a542fcd9.png)
I currently get a .tiff file with a single alpha channel using the code in my original post.
Example of what I currently output:
https://we.tl/t-fu3LcTbNnd
I hope this gives enough context

**Reply by LuukOV** [reply]
I also realized there's another module which allows more updated versions of ImageMagick: https://github.com/aheckmann/gm
Generally I'm fairly open to trying different software/versions as long as I can get the desired result.

**Comment by snibgo** [comment]
I know nothing about JS. I will show Windows syntax commands, and will leave any translation to JS to you.

The ImageMagick version number is something like 6.9.9-10 or 7.1.0-49. I suppose your "IM 0.1.3" is something different. What is the complete text output from "magick -version"?

I have named your input file `inSample.png`.

Copying alpha to a meta channel, and writing the result to a TIFF, should be simple and fast:
```
magick inSample.png -channel-fx "alpha=>meta" out.tiff
```
Sadly, this doesn't create a meta channel, so out.tiff contains just RGBA. I don't know why there is a problem. Perhaps there is a bug in IM, or in my brain.

The job can be done via "ftxt:", but it is slow and complex.

First, we write pixel values to text files:
```
magick ^
  inSample.png ^
  -precision 15 ^
  -write info: ^
  -channel RGBA ^
  -write ftxt:out1.txt ^
  -channel A ^
  -define ftxt:format="\v\n" ^
  -write ftxt:out2.txt ^
  NULL:
```
The "-write info:" tells us the size:
```
inSample.png PNG 4096x4096 4096x4096+0+0 8-bit sRGB 80096B 0.000u 0:00.000
```
We need the size later.

The first few lines of out1.txt are:
```
0,0:65535,65535,65535,0
1,0:65535,65535,65535,0
2,0:65535,65535,65535,0
3,0:65535,65535,65535,0
```
It also has lines like this:
```
1725,2656:0,0,0,65535
1726,2656:0,0,0,65535
1727,2656:0,0,0,65535
1728,2656:0,0,0,65535
```
I use Windows, so newlines are "\r\n". This would cause problems for the `paste` program, so I remove "\r" characters. Perhaps you can skip this step.
```
tr -d \r <out1.txt >out1a.txt
tr -d \r <out2.txt >out2a.txt
```
Now we use the `paste` program to join the lines together, separating the left and right sides by a comma "," character:
```
paste --delimiter=, out1a.txt out2a.txt >out3.txt
```
The first few lines of out3.txt are:
```
0,0:65535,65535,65535,0,0
1,0:65535,65535,65535,0,0
2,0:65535,65535,65535,0,0
3,0:65535,65535,65535,0,0
```
It has 5 channels, so we can read it with IM, and write a TIFF, as I showed above. Note the size must match what we found earlier.
```
magick -size 4096x4096 -define ftxt:hasalpha=true -define ftxt:nummeta=1 ftxt:out3.txt x.tiff
```
tiffinfo confirms that x.tiff has:
```
  Extra Samples: 2<unassoc-alpha, unspecified>
```
Note: tr, paste and tiffinfo are utilities commonly available, eg from Cygwin for Windows.

**Reply by LuukOV** [reply]
Thanks a lot! After a lot of fiddling, I managed to export an image containing 2 alpha channels, as I wanted. I however ran into another, separate issue. The resulting file I'm trying to get requires a clipping path and a certain colour-space. I've already managed to apply this through Gimp, however when I export it through gimp, the 2nd alpha channel is removed. I came up with the idea of maybe merging the two files somehow however I haven't found a way to do this yet. I mostly need to import the clipping path from one file to the other with the multiple alpha channels. Would you know any way to get the result I want by chance?
I can also confirm now that the version of IM I use is: (7.1.0-53)

**Comment by snibgo** [comment]
Note that a bug in "`-channel-fx`" has been fixed in the last day or two. See https://github.com/ImageMagick/ImageMagick/issues/5843 . This might mean that ...
```
magick inSample.png -channel-fx "alpha=>meta" out.tiff
```
... will now work.

> The resulting file I'm trying to get requires a clipping path ...

Do you want the clipping path in the result as a metadata vector path, or as an alpha channel?

IM can copy metadata vector path from one file to another. If you are combining two or more files in a "`magick`" command, metadata will come from the *first* file.

If you want the clipping path as an alpha channel, then that is simple -- you are simply combining raster images.

> I came up with the idea of maybe merging the two files somehow however I haven't found a way to do this yet.

I suggest you link to the input files and explain what result you want.

**Reply by LuukOV** [reply]
Doing the before mentioned command adds 5 fully black alpha channels for me:
https://we.tl/t-Y2AgzB0lmJ
to create this I used a file with the same setup as:
https://we.tl/t-fu3LcTbNnd

I meant the metadata vector path yeah. just as the above two files have. 
I've tried combining 1 file with a vector path and 1 file with the correct alpha channel setup (2 channels). but depending on the order, the image it outputs has either the path or the channels, but never both. it also blends the RGB colors but that's probably avoidable. 
The is the type of end-result I'm looking for:
https://we.tl/t-IG0fdeBIZY

**Comment by snibgo** [comment]
> to create this I used a file with the same setup as: https://we.tl/t-fu3LcTbNnd

The file there is named `example-current-output.tif`. It has an alpha channel and a clipping path. We can create a new tiff with the same alpha and path:
```
magick example-current-output.tif x.tiff
```
> I've tried combining 1 files with a vector path and 1 file with the correct alpha channel setup.

You want an output with the path from `example-current-output.tif` but alpha from another file, eg `getThisAlpha.tiff`. It should ignore colours from `getThisAlpha.tiff`. The output should have colours from `example-current-output.tif`. Is that right? I expect this should do what you want:
```
magick example-current-output.tif getThisAlpha.tiff -compose CopyOpacity -composite x2.tiff
```

**Reply by LuukOV** [reply]
I want the file "getThisAlpha.tiff" to have the two alpha channels like how we generate them as mentioned in the earlier comments. I've tried to do the mentioned command but the 2nd alpha channel gets stripped from the resulting .tiff sadly enough.

**Comment by LuukOV** [comment]
I'm still stuck on this issue, so far we've managed to generate a .tiff file with 2 alpha channels (showing up correctly in photoshop).
But I'm still not able to add a Path (clipping path) to the image. So I want to either append an extra alpha channel to an existing image which I already have that has a path or to add a path from another image to the image with the 2 alpha channels. Could anyone help me out?

**Comment by snibgo** [comment]
What are your input files? Which contains the clipping path? What command have you tried?

**Comment by LuukOV** [comment]
[Downloads.zip](https://github.com/ImageMagick/ImageMagick/files/10217650/Downloads.zip)
These are the files I have as input/output, also containing an example of the desired output I want. 
I've tried numerous commands so it's a bit hard to list them all here.

**Comment by snibgo** [comment]
`desired-output.tif` contains a path. I assume this is the output you want.

None of the other files contain a path.

**Comment by LuukOV** [comment]
Yes that's the output I want. It seems one of the files lost it's path. so this it my current output, containing only a single alpha channel but containing a path.
[output-single-alpha.zip](https://github.com/ImageMagick/ImageMagick/files/10218027/output-single-alpha.zip)

**Comment by snibgo** [comment]
Okay, `output-single-alpha.tif[0]` (note only one "F" in "tif") contains a clipping path. And you want that path to be in the output file. It has 4096x4096 pixels, including an alpha channel, but you don't want any pixel data from that image in your output. Is that correct?

What else do you want in your output file?

Your `desired-output.tif` contains two images: 4096x4096, and 2183x2137 pixels. I assume you want just the first image, which contains colour channels, and a clipping path, and alpha channel and a meta channel. The alpha and meta channels contain different data. Where should that colour, alpha and meta data come from?

**Comment by LuukOV** [comment]
Generally it's just a duplicate from the alpha channel that already exists (see `output-single-alpha.tif[1]`). 

Imagine as if I would duplicate the Alpha channel of `output-single-alpha.tif` to a meta channel within the same file. I create the path in the image using Gimp, gimp however doesn't save the meta channel. 

So I have to either move the path from the gimp exported image to the image with the alpha AND the meta channel or copy the Alpha channel of the image with the path to it's meta channel. 

I've tried the thing you mentioned before like:
` -channel-fx "alpha=>meta"`
But this didn't work the way I want it to, like I commented above. 

I also don't see the 2183x2137 image you mentioned in `desired-output.tif`. but eitherway I meant the 4096x4096 image yes.

**Comment by snibgo** [comment]
This post addresses "`-channel-fx`" only.

`toes_holed.png` is:
![toes_holed](https://user-images.githubusercontent.com/33812368/207393975-8c97552c-fda9-447a-9f37-b7cafc781009.png)

Using IM v7.1.0-54 downloaded today, "`-channel-fx alpha=>meta`" kind-of works, sort-of. Where `toes_holed.png` has an alpha channel, we try to copy that to a new "meta" channel:
```
magick toes_holed.png -channel-fx "alpha=>meta" x.tiff
```
Exiftool says:
```
exiftool x.tiff

:
Bits Per Sample                 : 8 8 8 8 8 8 8 8 8
:
Extra Samples                   : Unknown (2 0 0 0 0 0)
```
So x.tiff has nine channels. Oops. "`magick x.tiff -verbose info:`" says:
```
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
    Alpha: 1-bit
    Meta channel[0]: 8-bit
    Meta channel[1]: 8-bit
    Meta channel[2]: 8-bit
    Meta channel[3]: 8-bit
    Meta channel[4]: 8-bit
```
Hmm. We asked for one meta channel but have five. The meta channels seem to contain the same data as each other, but different to the alpha channel. We get the same result when we directly ...
```
magick toes_holed.png -channel-fx "alpha=>meta" -verbose info:
```
So IM still doesn't process `-channel-fx "alpha=>meta"` correctly. This is either a bug or "not yet implemented".

We get a similar problem with `-channel-fx "meta=0.5"`.

So, to copy the alpha channel into a new meta channel, I suggest the "ftxt:" method I detailed above.

(Incidentally, the IM language needs a mechanism to specify which meta channel we want, in "`-channel-fx`" and "`-channel`" and "`fx`".)

**Comment by LuukOV** [comment]
I've ran into this issue trying to copy to the meta channel as well, maybe something to report as a possible bug? Copying the alpha channel using ftxt works but I don't know yet how to keep the additional data like the clipping-path while doing that, hence my problem sadly enough.

**Comment by snibgo** [comment, marked as answer]
To copy alpha to a new meta channel, this Windows BAT script (named `alpha2meta.bat`) does the job:
```
rem %1 input image with an alpha channel
rem %2 output TIFF file, same as %1 but with a meta channel.

rem This would be fast and simple with
rem   -channel-fx alpha=>meta
rem but that doesn't currently (December 2022) work.

set INFILE=%1
set OUTFILE=%2

set TMPTXT1=out1.txt
set TMPTXT2=out2.txt
set TMPTXT3=out1a.txt
set TMPTXT4=out2a.txt
set TMPTXT5=out3.txt


for /F "usebackq" %%L in (`%IMG7%magick ^
  %INFILE%
  -precision 15 ^
  -format "WW=%%w\nHH=%%h\n" ^
  -write info: ^
  -channel RGBA ^
  -write ftxt:%TMPTXT1% ^
  -channel A ^
  -define "ftxt:format=\v\n" ^
  -write ftxt:%TMPTXT2% ^
  NULL:`) do set %%L

rem Windows uses \r\n as new lines, but paste needs \n only.
rem Strip "\r" from the text files.

tr -d \r <%TMPTXT1% >%TMPTXT3%
tr -d \r <%TMPTXT2% >%TMPTXT4%

paste --delimiter=, %TMPTXT3% %TMPTXT4% >%TMPTXT5%

%IMG7%magick ^
  -size %WW%x%HH% ^
  -define ftxt:hasalpha=true ^
  -define ftxt:nummeta=1 ^
  ftxt:%TMPTXT5% ^
  %OUTFILE%

exiftool %OUTFILE%
```

To copy all the Photoshop paths from one file (eg in.jpg) to another (eg out.tiff), where the output file already exists:
```
exiftool -config photoshop_paths.config -tagsfromfile in.jpg -allpaths# -originpathinfo out.tiff
```
The file `photoshop_paths.config` must exist in your current directory, or you must give the path to that file.

`photoshop_paths.config` can be downloaded from https://raw.githubusercontent.com/exiftool/exiftool/master/config_files/photoshop_paths.config

**Comment by LuukOV** [comment]
Perfect, thanks a lot! Using the exiftool command I managed to create an image having 2 alpha channels and the path I needed. I only have 1 small issue remaining. 

Currently the ExtraSamples property gets denoted as `(2 0)`. I somehow need to change this order to `(0 2)`. I think that means that currently it uses the 4th channel as pre-multiplied alpha (associated) and the 5th channel as a separate alpha (unspecified) channel. 

I've tried swapping the channels using `-channel-fx alpha<=>meta` but this doesn't seem to do what I want. 
The reason as to why I need this, is that currently if you open one of these files in photoshop it'll call the "seperate" alpha channel as "Alpha 2", I need it to be "Alpha 1".
[output-alpha2.zip](https://github.com/ImageMagick/ImageMagick/files/10246147/output-alpha2.zip)

**Comment by snibgo** [comment]
> Currently the ExtraSamples property gets denoted as (2 0). I somehow need to change this order to (0 2).

I don't think that is possible in IM. The first "ExtraSample" is what IM calls "alpha". That can be set to one of:

0 = Unspecified data
1 = Associated alpha data (with pre-multiplied color)
2 = Unassociated alpha data

For example, use "`-define tiff:alpha=unspecified`" before writing the tiff.

The second "ExtraSample" is what IM calls a meta-channel. IM has no mechanism for setting a meta-channel to unspecified, associated or unassociated.

**Comment by LuukOV** [comment]
Thank you for the quick reply, you've already helped a lot by letting me get this result so far. It's a shame this seems to be at the limit of IM, I imagine the FTXT might've allowed to save channel data in a prespecified order or the sort. Would you by chance know any other software I could look into to achieve this goal?

**Comment by snibgo** [comment]
When IM reads a `ftxt:` file, for each pixel it stores channel values in the same order that IM stores the channels in memory. That order is normally colour channels (R, G, B or C, M, Y, K) then alpha then meta-channels. I think that internally the order could be different (eg R, alpha, G, meta0, B, meta1, or R, G, B, meta0, alpha). However, there is no mechanism to change the order, as far as I know.

> Would you by chance know any other software I could look into to achieve this goal?

You might look at G'MIC (http://gmic.eu/), where "channel" is just another dimension, like "width" and "height". It is very easy to add new channels to an image. Thousands of channels if you want (and have enough memory). But I don't know if G'MIC can write those channels to a TIFF in the way you want.

**Reply by fmw42** [reply]
Can't you use -channel-fx to swap the order? See https://imagemagick.org/script/porting.php#cli

**Comment by snibgo** [comment]
If I understand correctly, "`-channel-fx red<=>blue`" should swap the values in those channels, rather than swapping the order of the channels in memory.

**Reply by fmw42** [reply]
OK. Thanks. I was not sure.

**Comment by LuukOV** [comment]
I figured out how to have the alpha layers in the way I want it.

I create a "template" file manually in photoshop, in this file I create the channels in the way I want it (transparency in the RGB and a separate alpha channel named "Alpha 1". I export this while having "save transparency" ticked. This will generate a file holding the photoshop tags I need (Alpha Channel Names). So instead of trying to add the tags later someway, I figured out I can use photoshop to make one and in a way copy them to my end-file.

Then in my processing process, I move the channel data of the 5 channel image I generated using the FTXT code you gave me above to a copy of my previously created template file. Then after this I move my path from another image to put it all together.

In the end I had to use a combination of GIMP, ImageMagick and ExifTool. Thank you for the continued support @snibgo, I managed to get the result I needed due to a big part of your help.
